Userland/Libraries/LibGfx/ImageFormats/AVIFLoader.cpp

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2023, Nico Weber <thakis@chromium.org>
+ * Copyright (c) 2024, doctortheemh <doctortheemh@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/ByteBuffer.h>
+#include <AK/Error.h>
+#include <LibGfx/ImageFormats/AVIFLoader.h>
+
+#include <avif/avif.h>
+
+namespace Gfx {
+
+class AVIFLoadingContext {
+    AK_MAKE_NONCOPYABLE(AVIFLoadingContext);
+
+public:
+    enum State {
+        NotDecoded = 0,
+        Error,
+        HeaderDecoded,
+        BitmapDecoded,
+    };
+
+    State state { State::NotDecoded };
+    ReadonlyBytes data;
+
+    avifDecoder* decoder { nullptr };
+
+    // image properties
+    Optional<IntSize> size;
+    bool has_alpha { false };
+    size_t image_count { 0 };
+    size_t repetition_count { 0 };
+    ByteBuffer icc_data;
+
+    Vector<ImageFrameDescriptor> frame_descriptors;
+
+    AVIFLoadingContext() = default;
+    ~AVIFLoadingContext()
+    {
+        avifDecoderDestroy(decoder);
+        decoder = nullptr;
+    }
+};
+
+AVIFImageDecoderPlugin::AVIFImageDecoderPlugin(ReadonlyBytes data, OwnPtr<AVIFLoadingContext> context)
+    : m_context(move(context))
+{
+    m_context->data = data;
+}
+
+AVIFImageDecoderPlugin::~AVIFImageDecoderPlugin()
+{
+}
+
+static ErrorOr<void> decode_avif_header(AVIFLoadingContext& context)
+{
+    if (context.state >= AVIFLoadingContext::HeaderDecoded)
+        return {};
+
+    if (context.decoder == nullptr) {
+        context.decoder = avifDecoderCreate();
+
+        if (context.decoder == nullptr) {
+            return Error::from_string_literal("failed to allocate AVIF decoder");
+        }
+    }
+
+    avifResult result = avifDecoderReset(context.decoder);
+    if (result != AVIF_RESULT_OK)
+        return Error::from_string_literal("Failed to reset decoder");
```

Comment:
This doesn't seem to be useful/mandatory.
https://github.com/AOMediaCodec/libavif/blob/b0eb897c44c27a06e91f1326eba7cb1f0eb2c8a5/include/avif/avif.h#L1330-L1347

Reply:
Agreed after further review of the header. We invoke Parse immediately.